.werft/platform-delete-preview-environments-cron.ts

```diff
@@ -37,34 +41,98 @@ async function deletePreviewEnvironments() {
-    // preview name patterns before it is safe to delete a namespace.
-    const expectedPreviewEnvironmentNamespaces = new Set(branches.flatMap(branch => [parseBranch(branch), expectedNamespaceFromBranch(branch)]));
-    werft.done("Fetching branches");
-
     werft.phase("Fetching previews");
     let previews: string[]
     try {
         previews = listAllPreviewNamespaces(CORE_DEV_KUBECONFIG_PATH, {});
-        previews.forEach(previewNs => werft.log("Fetching previews", previewNs))
+        previews.forEach(previewNs => werft.log("Fetching preview", previewNs));
+        werft.done("Fetching preview");
     } catch (err) {
-        werft.fail("Fetching previews", err)
+        werft.fail("Fetching preview", err)
     }
-    werft.done("Fetching previews");
+
+    werft.phase("Fetching outdated branches");
+    const branches = getAllBranches();
+    const outdatedPreviews = new Set(branches
+        .filter(branch => {
+          const lastCommit = exec(`git log origin/${branch} --since=$(date +%Y-%m-%d -d "5 days ago")`, { silent: true })
+          return lastCommit.length < 1
+        })
+        .map(branch => expectedNamespaceFromBranch(branch)))
+
+    const expectedPreviewEnvironmentNamespaces = new Set(branches.map(branch => expectedNamespaceFromBranch(branch)));
 
     werft.phase("deleting previews")
     try {
-        const previewsToDelete = previews.filter(ns => !expectedPreviewEnvironmentNamespaces.has(ns))
-        // Trigger namespace deletion in parallel
-        const promises = previewsToDelete.map(preview => wipePreviewEnvironmentAndNamespace(helmInstallName, preview, CORE_DEV_KUBECONFIG_PATH, { slice: `Deleting preview ${preview}` }));
-        // But wait for all of them to finish before (or one of them to fail) before we continue
-        await Promise.all(promises)
+        const deleteDueToMissingBranch     = previews.filter(ns => !expectedPreviewEnvironmentNamespaces.has(ns))
+        const deleteDueToNoCommitActivity  = previews.filter(ns => outdatedPreviews.has(ns))
+        const deleteDueToNoDBActivity      = previews.filter(ns => isInactive(ns))
+        const previewsToDelete             = new Set([...deleteDueToMissingBranch, ...deleteDueToNoCommitActivity, ...deleteDueToNoDBActivity])
+
+        if (previewsToDelete.has("staging-main")) {
+            previewsToDelete.delete("staging-main")
+        }
+
+        if (DRY_RUN) {
+            previewsToDelete.forEach(preview => werft.log("deleting preview", `would have deleted preview environment ${preview}`))
+        }
+        else {
+            const promises: Promise<any>[] = [];
+            previewsToDelete.forEach(preview => {
+                werft.log("deleting preview", preview)
+                promises.push(wipePreviewEnvironmentAndNamespace(helmInstallName, preview, CORE_DEV_KUBECONFIG_PATH, { slice: `Deleting preview ${preview}` }))
+            })
+            await Promise.all(promises)
```

Comment:
I had to change the mapping as the compiler throw an arrow that `.map` is not supported on this set.